### PR TITLE
Update Intel x86 CPU presets to match LLVM

### PR DESCRIPTION
**Pull request by MozarellaMan**
This is for #5444 

I decided to just start with the Intel CPUs for now just to get feedback on whether this is an alright way to sync up + update cranelift's to match LLVM's presets. I used [this](https://github.com/llvm/llvm-project/blob/d4493dd1ed58ac3f1eab0c4ca6e363e2b15bfd1c/llvm/lib/Target/X86/X86.td#L1300-L1643) (thanks @bjorn3!) as a reference for the features and presets used by LLVM.

## Changes
- added `Copy, Clone` derive to `PresetIndex` to enable composing presets multiple times
- added missing x86 Intel CPU presets (old and new)
- updated existing presets with apparently missing features (`cannonlake` and `icelake`)

## Notes
- What's the best way to test this? Can it be meaningfully tested?
- Should I include links to the exact feature lists for each preset as a comment with each preset or just one link at the top of all the presets?
- In some cases, LLVM has a different name for the same architecture feature e.g. LLVM has `FeatureDQI`, cranelift has `has_avx512dq` - this led to me updating the existing presets with features that those archtectures have (or at least LLVM says they have those features!)

### CODE REVIEW

**Comment by bjorn3** [comment]
I think this looks good. Maybe add a comment at the top of the file referencing the exact file+commit you used? The github url you linked would work.

**Comment by jameysharp** [comment]
Thank you, at a quick glance this looks great! A bunch of people are out due to holidays, but if nobody else gets to it first I intend to review this next week. Feel free to comment again if you don't hear anything by the end of next week.

**Comment by jameysharp** [comment]
This seems like the right direction!

My only quibble is that I think we should only include 64-bit CPUs (the ones that have `FeatureX86_64` in LLVM). We should be able to generate code for any x86 CPU running in 64-bit mode, and if we can't right now I think we're at least close. But I believe we'd need to do a lot more work to support 32-bit mode, so I think it makes sense to just fail if someone requests a target that can only support 32-bit mode.

I haven't yet checked that, for each target we already supported, the set of features is still correct. I believe you probably got it right, but the LLVM tables are hard enough to read that I think it's worth double-checking, just in case.

Overall, I think this is the right patch. If you could remove those 32-bit CPUs I'd appreciate it, and hopefully I'll get to double-checking the features lists soon, or talk somebody else into doing it. :grin:

File: cranelift/codegen/meta/src/isa/x86.rs [hunk 1]
```diff
@@ -174,38 +174,200 @@ fn define_settings(shared: &SettingGroup) -> SettingGroup {
...
+    );
+    let sapphire_rapids = settings.add_preset(
+        "sapphirerapids",
+        "Saphire Rapids microarchitecture.",
+        preset!(icelake_server),
+    );
+    let alderlake = settings.add_preset(
+        "alderlake",
+        "Alder Lake microarchitecture.",
+        preset!(tremont && has_fma && has_bmi1 && has_bmi2 && has_lzcnt),
```
Comment (abrown) [inline cranelift/codegen/meta/src/isa/x86.rs:353]: Is this duplicated from up above?

**Review by jameysharp — COMMENTED** [review]
I've finished reviewing this. Nice work! There are a couple of issues but mostly I think this is exactly what I'd want it to look like.

**Comment by MozarellaMan** [comment]
Thank you for the comments! Sorry, I was away a bit. Will have a look at addressing what you've brought up. Thanks for taking the time to double check - I didn't realise the difference regarding the feature implication

File: cranelift/codegen/meta/src/isa/x86.rs [hunk 2]
```diff
@@ -174,38 +174,200 @@ fn define_settings(shared: &SettingGroup) -> SettingGroup {
...
+    // Intel Core Duo
+    settings.add_preset("yonah", "Yonah microarchitecture.", preset!(has_sse3));
+
+    // Netburst
+    settings.add_preset("prescott", "Prescott microarchitecture.", preset!(has_sse3));
+    settings.add_preset("nocona", "Nocona microarchitecture.", preset!(has_sse3));
+
+    // Intel Core 2 Solo/Duo
+    settings.add_preset("core2", "Core 2 microarchitecture.", preset!(has_ssse3));
+    settings.add_preset("penryn", "Penryn microarchitecture.", preset!(has_sse41));
```
[the remark this answers is not in the document]
Comment (MozarellaMan) [reply cranelift/codegen/meta/src/isa/x86.rs]: Personally, I would prefer the features implying the others - it makes sense now that you pointed it out. The added description in the setting preset makes that more obivous.

File: cranelift/codegen/meta/src/isa/x86.rs [hunk 3]
```diff
@@ -194,15 +194,15 @@ fn define_settings(shared: &SettingGroup) -> SettingGroup {
     // Intel Atom CPUs
     let atom = settings.add_preset("atom", "Atom microarchitecture.", preset!(has_ssse3));
     settings.add_preset("bonnell", "Bonnell microarchitecture.", preset!(atom));
-    let silvermont = settings.add_preset(
-        "silvermont",
+    let slm = settings.add_preset(
+        "slm",
```
Comment (jameysharp) [inline cranelift/codegen/meta/src/isa/x86.rs:194]: If I'm reading the LLVM tables correctly, it exposes this as _both_ "slm" and "silvermont", but defines them identically. So we should do the same. I think it makes sense to use the longer name "silvermont" as the let-bound preset, and define the "slm" preset in terms of "silvermont".

**Review by jameysharp — COMMENTED** [review]
It's no problem if you take some time getting through this—I really appreciate you working on it!

I have one clarification to make about the changes I suggested but you've addressed most of what I asked for. Thank you!

File: cranelift/codegen/meta/src/isa/x86.rs (hunk 1 above)
Comment (jameysharp) [reply cranelift/codegen/meta/src/isa/x86.rs:333]: I think the duplicate definition of `alderlake` is the last issue remaining here. Once that's fixed I think this PR is ready!

**Pull request merged**